app/src/main/java/org/mozilla/fenix/components/metrics/LeanplumMetricsService.kt

```diff
@@ -72,13 +77,30 @@ class LeanplumMetricsService(private val application: Application) : MetricsServ
     override val type = MetricServiceType.Marketing
     private val token = Token(LeanplumId, LeanplumToken)
 
+    private val preferences = application.getSharedPreferences(PREFERENCE_KEY, MODE_PRIVATE)
+
+    @VisibleForTesting
+    internal var deviceId: String
+        get() {
+            var deviceId = preferences.getString(DEVICE_ID_KEY, null)
+            if (deviceId == null) {
+                deviceId = deviceIdGenerator.invoke()
+                this.deviceId = deviceId
+            }
+
+            return deviceId
+        }
+        set(value) {
+            preferences.edit().putString(DEVICE_ID_KEY, value).apply()
+        }
```

Comment:
I wonder if we can turn this into a `val`/`by lazy` property? Not having to expose the setter may prevent errors in the future. :)